Added Web3 Staking Dapp

nirban256 opened the pull request:
# Description

It is a Staking Dapp where people will get interest based on the amount they stake after a specific period of time. I have added some dependencies namely hardhat and ethers in order to build the Dapp (Decentralized Application).

Fixes #550 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

# Checklist:

- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

nirban256 commented:
@atapas pls review the pr

atapas commented on src/plays/web3-staking-dapp/artifacts/build-info/5faf9c6c2f8753632f3c7c13937cd196.json:
```diff
@@ -0,0 +1 @@
+{"id":"5faf9c6c2f8753632f3c7c13937cd196","_format":"hh-sol-build-info-1","solcVersion":"0.8.0","solcLongVersion":"0.8.0+commit.c7dfd78e","input":{"language":"Solidity","sources":{"contracts/Staking.sol":{"content":"// SPDX-License-Identifier: UNLICENSED\npragma solidity ^0.8.0;\n\ncontract Staking {\n    address public admin;\n\n    struct Stakers {\n        uint256 stakerId;\n        address sta ...
```
Is this file intended?

nirban256 replied:
I think it will be better if we keep this file

atapas commented on src/plays/web3-staking-dapp/artifacts/contracts/Staking.sol/Staking.dbg.json:
```diff
@@ -0,0 +1,4 @@
+{
```
Is this file needed?

nirban256 replied:
I think for a web3 dapp we should keep all the files related to it so I have kept the files

atapas commented on src/plays/web3-staking-dapp/styles.css:
```diff
@@ -0,0 +1,214 @@
+/* Navbar */
```
Remove commented code

nirban256 replied:
ok

atapas reviewed (changes requested):
I have given some initial comments.

@Sachin-chaurasiya will do deep review.

Sachin-chaurasiya commented on src/plays/web3-staking-dapp/Readme.md:
```diff
@@ -0,0 +1,27 @@
+- Blog: null
+- Video: https://www.youtube.com/watch?v=AjaHJicbPkM&ab_channel=NirbanChakraborty
+
+## Implementation Details
+
+Update your implementation idea and details here
+
+## Consideration
+
+Update all considerations(if any)
+
+## Resources
+
+Update external resources(if any)
```
Can we update these details as well?

Sachin-chaurasiya commented on src/plays/web3-staking-dapp/Web3StakingDapp.js:
```diff
@@ -0,0 +1,273 @@
+
+  // Your Code Start below.
+
+  const ContractAddress = '0x12163B070B97f06F5061D93164D960bbFCfdf965';
+
```
Move this to constant.

nirban256 replied:
So I have to create a file called constant,js and then I have to keep the contract address there?
This is what you are saying

Sachin-chaurasiya replied:
Yes, Exactly.

nirban256 replied:
ok

Sachin-chaurasiya commented on src/plays/web3-staking-dapp/Web3StakingDapp.js:
```diff
@@ -0,0 +1,273 @@
+  const ContractAddress = '0x12163B070B97f06F5061D93164D960bbFCfdf965';
+
+  // frontend states
+  const [provider, setProvider] = useState(undefined);
+  const [account, setAccount] = useState(undefined);
+  const [contract, setContract] = useState(undefined);
+  const [accountAddress, setAccountAddress] = useState(undefined);
```
We don't need to provide the default value as `undefined` as it will be undefined by default.
```suggestion
  const [provider, setProvider] = useState();
  const [account, setAccount] = useState();
  const [contract, setContract] = useState();
  const [accountAddress, setAccountAddress] = useState();
```

Sachin-chaurasiya commented on src/plays/web3-staking-dapp/Web3StakingDapp.js:
```diff
@@ -0,0 +1,273 @@
+
+  // staking
+  const [stakeModal, setStakeModal] = useState(false);
+  const [stakingLength, setStakingLength] = useState(undefined);
+  const [stakingPercent, setStakingPercent] = useState(undefined);
```
```suggestion
  const [stakingLength, setStakingLength] = useState();
  const [stakingPercent, setStakingPercent] = useState();
```

Sachin-chaurasiya commented on src/plays/web3-staking-dapp/Web3StakingDapp.js:
```diff
@@ -0,0 +1,273 @@
+  const toMatic = wei => ethers.utils.formatEther(wei);
+
+  useEffect(() => {
+    const onload = async () => {
+      const provider = new ethers.providers.Web3Provider(window.ethereum);
+      setProvider(provider);
+
+      const contract = new ethers.Contract(
+        ContractAddress,
+        artifact.abi
+      )
+      setContract(contract);
+    }
```
Can we Move this function out of the useEffect?

nirban256 replied:
then it will show a warning i think

Sachin-chaurasiya replied:
Which warning?

Sachin-chaurasiya commented on src/plays/web3-staking-dapp/smart-contract/cache/solidity-files-cache.json:
```diff
@@ -0,0 +1,40 @@
+{
```
Can we not include the cache files?

nirban256 replied:
can you tell me the what to write in the gitignore in order to remove the cache file?
I have tried with ./smart-contract/cache/solidity-files-cache.json

Sachin-chaurasiya replied:
Can you try `cache/*`?

nirban256 replied:
it didn't worked

Sachin-chaurasiya replied:
@nirban256 , if you have already pushed the changes and they trying to ignore that files in gitignore then it won't work.

nirban256 replied:
@Sachin-chaurasiya I made another commit after editing the readme file and then it failed the test

nirban256 replied:
I don't know the exact reason for it

atapas replied:
I've re-run the failed test. It should pass now. Can you please take care of the comments making sure they are done?

nirban256 replied:
@atapas I could not add the cache folder in the gitignore properly this is the only issue that is left. Can you tell me how to add the folder so that git does not track it?

nirban256 replied:
ok

Sachin-chaurasiya reviewed (changes requested):
Some initial comments

nirban256 commented:
If all the conditions are met then please merge the pr

Sachin-chaurasiya reviewed (changes requested):
@nirban256 Few more comments and remove the `src/plays/web3-staking-dapp/styles.css` as it's empty.

nirban256 commented on src/plays/web3-staking-dapp/smart-contract/scripts/deploy.js:
```diff
@@ -0,0 +1,55 @@
+// We recommend this pattern to be able to use async/await everywhere
+// and properly handle errors.
+main().catch((error) => {
+  console.error(error);
```
having this statement will not create any problem because this file will not run when the play will run, it's work is done.

Sachin-chaurasiya commented:
@nirban256 , Build is failing.

<img width="1213" alt="image" src="https://user-images.githubusercontent.com/59080942/198881602-c97d6d5a-ef78-4f33-8614-b3e0aa79abfc.png">

nirban256 commented:
> @nirban256 , Build is failing, As we have deleted the styles.css and we have to remove the reference of it as well.
> 
> <img alt="image" width="1213" src="https://user-images.githubusercontent.com/59080942/198881602-c97d6d5a-ef78-4f33-8614-b3e0aa79abfc.png">

removed the line @Sachin-chaurasiya

Sachin-chaurasiya reviewed (approved):
Great work @nirban256

nirban256 commented:
thanks @Sachin-chaurasiya

atapas reviewed (approved):
Approving
